### ISSUE TITLE
Wrong colors in the preview image when using profile denoise (wavelets) under Windows

### ISSUE THREAD

**Issue by Mark-64** [darktable-org/darktable#7814]
Now this is another fantastic Windows only issue. It happen with and without OpenCL

After #7587 I get often wrong color when using profiled denoise wavelets in combination with few other basic modules.

See before the commit (a recent PlayRaw image)
![Cattura13](https://user-images.githubusercontent.com/43290988/104336374-a1e7ab80-54f4-11eb-95e1-9504943a9959.JPG)

and after
![Cattura12](https://user-images.githubusercontent.com/43290988/104336487-c04da700-54f4-11eb-8ce8-ea10c6fb849f.JPG)

Here are the raw file and the xmp

[test.zip](https://github.com/darktable-org/darktable/files/5803149/test.zip)

**Comment by MStraeten** [comment]
confirmed for  darktable 3.5.0+558~gd82d67fbe on win10
with opencl enabled it occurs only if the pixpelpipe is processed without opencl 
preview and full are fine if forced to be processed by opencl `opencl_device_priority=0,*/0,*/*/*/!0,*`

**Comment by TurboGit** [comment]
Worrisome ! @ralfbrown can you have look ?

**Comment by johnny-bit** [comment]
Duplicate of #7686

**Comment by Mark-64** [comment]
Yes but I bisect the issue ;-)
@ I believe it is related to compiled optimizations behaving weird in Windows, we had other cases.
Will try to recompile without optimizations

**Comment by ralfbrown** [comment]
> Worrisome ! @ralfbrown can you have look ?

Will do, but probably won't get to it until the weekend.

**Comment by TurboGit** [comment]
No problem there is no urgency, Thanks.

**Comment by Mark-64** [comment]
As expected, commenting out the `#ifdef _OPENMP` directives in /iop/denoiseprofile.c makes the problem disappear.
Now I have to understand which one is producing the issue

**Comment by ralfbrown** [comment]
> At times the issue disappears commenting out parts of the code which are not even executed!

Traditionally, that's pretty definitive evidence of using uninitialized memory (the code size changes, therefore the accessed locations change).  But on modern systems, the heap generally isn't contiguous with the code anymore, so there's not as direct a link.

Is it possible for you to compile with Address Sanitizer or to run Valgrind?

**Comment by Mark-64** [comment]
> Is it possible for you to compile with Address Sanitizer or to run Valgrind?

I've never used such tools, I will have to document myself, will let you know

**Comment by ralfbrown** [comment]
I haven't used Windows since ME :-), which is why I asked whether it's possible.

Address Sanitizer is part of GCC and Clang (though it relies on an optional library), and if available can be made part of the build with `./build.sh --asan`.  You can then just run the program and you'll get a report on stdout of any problems found.  I don't think Asan finds reads of uninitialized memory like Valgrind does, but it will find out of bounds accesses and references to just-freed memory.

Valgrind doesn't require recompilation, you just run the program under valgrind (but it runs 10-20 times slower because it actually interprets the machine code, and while it supports multiple threads, it only uses one core, which slows things down even further).

**Comment by Mark-64** [comment]
Well, for the moment I managed to compile and run with ASAN in Ubuntu (later will try on Windows too).
Starting from and empty config and trying to import the one single image I posted above, before even reaching the thumbnail display in lighttable the program halts with the following output
```
marco@marco-desktop:~/develop/darktable/build/Asan$ /opt/darktable/bin/darktable
[defaults] found a 64-bit system with 16321556 kb ram and 16 cores (0 atom based)
[defaults] setting very high quality defaults
=================================================================
==25955==ERROR: AddressSanitizer: new-delete-type-mismatch on 0x603000ad4460 in thread T0:
  object passed to delete has wrong type:
  size of the allocated type:   32 bytes;
  size of the deallocated type: 24 bytes.
    #0 0x7f203f230025 in operator delete(void*, unsigned long) (/usr/lib/x86_64-linux-gnu/libasan.so.5+0x111025)
    #1 0x7f202819e704  (/lib/x86_64-linux-gnu/liblensfun.so.1+0x14704)
    #2 0x7f202819f040 in lfModifier::~lfModifier() (/lib/x86_64-linux-gnu/liblensfun.so.1+0x15040)
    #3 0x7f20281cbaa6 in get_autoscale /home/marco/develop/darktable/src/iop/lens.cc:2162
    #4 0x7f20281cf00e in reload_defaults /home/marco/develop/darktable/src/iop/lens.cc:1383
    #5 0x7f203ee8868a in dt_iop_reload_defaults /home/marco/develop/darktable/src/develop/imageop.c:1452
    #6 0x7f203ee7ffc7 in _dt_dev_load_pipeline_defaults /home/marco/develop/darktable/src/develop/develop.c:644
    #7 0x7f203ee7ffc7 in dt_dev_read_history_ext /home/marco/develop/darktable/src/develop/develop.c:1733
    #8 0x7f203ee7ffc7 in dt_dev_read_history_ext /home/marco/develop/darktable/src/develop/develop.c:1710
    #9 0x7f203ee8121a in dt_dev_load_image /home/marco/develop/darktable/src/develop/develop.c:725
    #10 0x7f202893de88 in enter /home/marco/develop/darktable/src/views/darkroom.c:2933
    #11 0x7f203ef89e71 in dt_view_manager_switch_by_view /home/marco/develop/darktable/src/views/view.c:451
    #12 0x7f203edfb641 in _dt_ctl_switch_mode_to /home/marco/develop/darktable/src/control/control.c:356
    #13 0x7f203e7c8084 in g_main_context_invoke_full (/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x53084)
    #14 0x7f2022c9e0e5 in _lib_import_single_image_callback /home/marco/develop/darktable/src/libs/import.c:566
    #15 0x7f203dae4a55  (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x14a55)
    #16 0x7f203db03b27 in g_signal_emit_valist (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x33b27)
    #17 0x7f203db040d2 in g_signal_emit (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x340d2)
    #18 0x7f203e1022ad  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x13e2ad)
    #19 0x7f203e102317  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x13e317)
    #20 0x7f203dae4801 in g_closure_invoke (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x14801)
    #21 0x7f203daf8961  (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x28961)
    #22 0x7f203db03b9d in g_signal_emit_valist (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x33b9d)
    #23 0x7f203db040d2 in g_signal_emit (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x340d2)
    #24 0x7f203e100753  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x13c753)
    #25 0x7f203e3b1ae0  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x3edae0)
    #26 0x7f203dae4a55  (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x14a55)
    #27 0x7f203db03b27 in g_signal_emit_valist (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x33b27)
    #28 0x7f203db040d2 in g_signal_emit (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x340d2)
    #29 0x7f203e1c8fcb  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x204fcb)
    #30 0x7f203dae7c55 in g_cclosure_marshal_VOID__BOXEDv (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x17c55)
    #31 0x7f203dae4a55  (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x14a55)
    #32 0x7f203db03b27 in g_signal_emit_valist (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x33b27)
    #33 0x7f203db040d2 in g_signal_emit (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x340d2)
    #34 0x7f203e1c6011  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x202011)
    #35 0x7f203e1c765a  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x20365a)
    #36 0x7f203e1ca645  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x206645)
    #37 0x7f203e191baf in gtk_event_controller_handle_event (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x1cdbaf)
    #38 0x7f203e35416c  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x39016c)
    #39 0x7f203e3ab5ee  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x3e75ee)
    #40 0x7f203dae4a55  (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x14a55)
    #41 0x7f203db02dd0 in g_signal_emit_valist (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x32dd0)
    #42 0x7f203db040d2 in g_signal_emit (/lib/x86_64-linux-gnu/libgobject-2.0.so.0+0x340d2)
    #43 0x7f203e355c22  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x391c22)
    #44 0x7f203e211127  (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x24d127)
    #45 0x7f203e2133da in gtk_main_do_event (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x24f3da)
    #46 0x7f203defbf78  (/lib/x86_64-linux-gnu/libgdk-3.so.0+0x3cf78)
    #47 0x7f203df2f105  (/lib/x86_64-linux-gnu/libgdk-3.so.0+0x70105)
    #48 0x7f203e7c6fbc in g_main_context_dispatch (/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x51fbc)
    #49 0x7f203e7c723f  (/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x5223f)
    #50 0x7f203e7c7532 in g_main_loop_run (/lib/x86_64-linux-gnu/libglib-2.0.so.0+0x52532)
    #51 0x7f203e21237c in gtk_main (/lib/x86_64-linux-gnu/libgtk-3.so.0+0x24e37c)
    #52 0x7f203ef293bc in dt_gui_gtk_run /home/marco/develop/darktable/src/gui/gtk.c:1551
    #53 0x56542533c0d5 in main /home/marco/develop/darktable/src/main.c:93
    #54 0x7f203eaa70b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #55 0x56542533c10d in _start (/opt/darktable/bin/darktable+0x110d)

0x603000ad4460 is located 0 bytes inside of 32-byte region [0x603000ad4460,0x603000ad4480)
allocated by thread T0 here:
    #0 0x7f203f22e947 in operator new(unsigned long) (/usr/lib/x86_64-linux-gnu/libasan.so.5+0x10f947)
    #1 0x7f202819e0c4 in lfModifier::AddSubpixelCallback(void (*)(void*, float*, int), int, void*, unsigned long) (/lib/x86_64-linux-gnu/liblensfun.so.1+0x140c4)

SUMMARY: AddressSanitizer: new-delete-type-mismatch (/usr/lib/x86_64-linux-gnu/libasan.so.5+0x111025) in operator delete(void*, unsigned long)
==25955==HINT: if you don't care about these errors you may set ASAN_OPTIONS=new_delete_type_mismatch=0
==25955==ABORTING
```

**Comment by ralfbrown** [comment]
That's in the lensfun code rather than darktable proper.  Do as instructed to ignore that class of errors.
Forgot to mention that little "gotcha"....

**Comment by Mark-64** [comment]
[valgrind.log](https://github.com/darktable-org/darktable/files/5813236/valgrind.log)
[asan.log](https://github.com/darktable-org/darktable/files/5813239/asan.log)
I can't find useful info in those... now trying in Windows

**Comment by Mark-64** [comment]
... and in Windows none of the tools suggested work, unforunately

**Comment by Mark-64** [comment]
update: on windows i can't use the dignostic tools but I found that reverting all the changes made in #7587 in `process_wavelets()` and deleting `variance_stabilizing_xform()` which is no longer used in that case, the problem seems to be solved reliably.
So the problem should be in one of these two functions.

**Comment by ralfbrown** [comment]
> update: on windows i can't use the dignostic tools but I found that reverting all the changes made in #7587 in `process_wavelets()` and deleting `variance_stabilizing_xform()` which is no longer used in that case, the problem seems to be solved reliably.

That's effectively completely reverting the PR :-).

You were right that there wasn't any useful info in the Valgrind and Asan logs.  Some uninit reads to be checked into, but completely unrelated to the denoising code.

Unfortunately your reproducer doesn't on my system, building either as Release or RelWithDebInfo, so it's going to be a matter of blindly trying various things.  One thing that may help narrow down the problem: do you get the green output if you use RGB instead of Y0U0V0?  There's some use of white balance factors that looks suspicious in the YUV normalization path (specifically in the backtransform).

**Comment by Mark-64** [comment]
> do you get the green output if you use RGB instead of Y0U0V0?

No, it happens exclusively with wevelets and Y0U0V0. 
As you can't replicate the issue, don't hesitate to give me things to test.

**Comment by ralfbrown** [comment]
Please give #7900 a try.  Hopefully one of the changes there will manage to work around the issue.  I did find that zapping one of the channels before doing the Y0U0V0 back-transform yielded a very green (but not to the extent shown up-thread) result, so maybe rearranging the memory layout of things as part of the vectorizability changes in 7900 will work around whatever was getting messed up.
